consensus/src/consensus.rs

```diff
@@ -0,0 +1,174 @@
+pub struct ConsensusState<PublicKey: VerifyingKey> {
+    /// The last committed round.
+    pub last_committed_round: Round,
+    // Keeps the last committed round for each authority. This map is used to clean up the dag and
+    // ensure we don't commit twice the same certificate.
+    pub last_committed: HashMap<PublicKey, Round>,
+    /// Keeps the latest committed certificate (and its parents) for every authority. Anything older
+    /// must be regularly cleaned up through the function `update`.
+    pub dag: Dag<PublicKey>,
+}
+
+impl<PublicKey: VerifyingKey> ConsensusState<PublicKey> {
+    pub fn new(genesis: Vec<Certificate<PublicKey>>) -> Self {
+        let genesis = genesis
+            .into_iter()
+            .map(|x| (x.origin(), (x.digest(), x)))
+            .collect::<HashMap<_, _>>();
+
+        Self {
+            last_committed_round: 0,
+            last_committed: genesis
+                .iter()
+                .map(|(x, (_, y))| (x.clone(), y.round()))
+                .collect(),
+            dag: [(0, genesis)]
+                .iter()
+                .cloned()
+                .collect::<HashMap<_, HashMap<_, _>>>(),
+        }
+    }
+
+    /// Update and clean up internal state base on committed certificates.
+    pub fn update(&mut self, certificate: &Certificate<PublicKey>, gc_depth: Round) {
+        self.last_committed
+            .entry(certificate.origin())
+            .and_modify(|r| *r = max(*r, certificate.round()))
+            .or_insert_with(|| certificate.round());
+
+        let last_committed_round = *std::iter::Iterator::max(self.last_committed.values()).unwrap();
+        self.last_committed_round = last_committed_round;
+
+        // We purge all certificates past the gc depth
+        self.dag.retain(|r, _| r + gc_depth >= last_committed_round);
+        for (name, round) in &self.last_committed {
+            self.dag.retain(|r, authorities| {
+                // We purge certificates for `name` prior to its latest commit
+                if r < round {
+                    authorities.retain(|n, _| n != name);
+                }
+                !authorities.is_empty()
+            });
+        }
+    }
+}
+
+/// Describe how to sequence input certificates.
+pub trait ConsensusProtocol<PublicKey: VerifyingKey> {
+    fn process_certificate(
+        &mut self,
+        // The state of the consensus protocol.
+        state: &mut ConsensusState<PublicKey>,
+        // The latest consensus index.
+        consensus_index: SequenceNumber,
+        // The new certificate.
+        certificate: Certificate<PublicKey>,
+    ) -> StoreResult<Vec<ConsensusOutput<PublicKey>>>;
+}
+
+pub struct Consensus<PublicKey: VerifyingKey, ConsensusProtocol> {
+    /// Receives new certificates from the primary. The primary should send us new certificates only
+    /// if it already sent us its whole history.
+    rx_primary: Receiver<Certificate<PublicKey>>,
```

Comment:
Have we ever considered output certificates before synchronizing them? Especially if we have an ordering machine and a separate execution machine in the future it might be worth not following our old convention

Reply:
Yes we disscussed the for Celo and Somelier but didn't end up in our current roadmap. For Sui it is less important since "execution" simply means incrementing a sequence number (so it is very fast)